### PR TITLE
fix: typeahead query timeout

### LINKED ISSUE #6712: ActiveRecord::QueryCanceled: PG::QueryCanceled: ERROR:  canceling statement due to statement timeout (ActiveRecord::QueryCanceled)
Sentry Issue: [CIRCUITVERSE-CORE-Y7](https://circuitverse.sentry.io/issues/90256698/)

```
PG::QueryCanceled: ERROR:  canceling statement due to statement timeout (PG::QueryCanceled)
  app/controllers/users/circuitverse_controller.rb:26:in `typeahead_educational_institute'
    .pluck(:educational_institute)
  app/controllers/application_controller.rb:35:in `switch_locale'
    I18n.with_locale(locale, &)
...
(131 additional frame(s) were not displayed)

ActiveRecord::QueryCanceled: PG::QueryCanceled: ERROR:  canceling statement due to statement timeout (ActiveRecord::QueryCanceled)
```
This issue was automatically created by Sentry via [New issue](https://sentry.io/organizations/circuitverse/monitors/alerts/12755/)

### PR DESCRIPTION
**Pull request by ayuxsh009**
Fixes #6712

#### Describe the changes you have made in this PR -
This PR fixes frequent `PG::QueryCanceled` timeouts in the `typeahead_educational_institute` endpoint by optimizing the search query and adding safeguards to prevent expensive requests from reaching the database.

The previous implementation used `LIKE '%query%'`, which caused full table scans and resulted in slow queries under load. This PR switches the endpoint to use the existing GIN-indexed full-text search column, adds input validation + caching, and gracefully handles query timeouts.

### Screenshots of the UI changes (If any) -
N/A (No UI changes)

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**Explain your implementation approach:**
This PR solves the timeout issue in the institute typeahead by replacing an expensive wildcard `LIKE` search with indexed full-text search.

**Alternative approaches considered:**
- Adding/optimizing indexes for `LIKE` queries (not effective with leading `%`)
- Increasing DB statement timeout (does not solve the root cause)
- Only limiting result size (query would still remain costly)

**Why this implementation was chosen:**
The project already maintains a GIN-indexed searchable `tsvector` column. Using full-text search allows Postgres to leverage the index properly, making the endpoint fast and stable.

**Key changes included:**
- Updated query logic to use the existing full-text searchable `tsvector` column
- Added input validation (minimum 2 characters + sanitization) to avoid unnecessary heavy queries
- Added 5-minute caching to reduce repeated DB hits for common searches
- Added safe handling for canceled/timeout queries by returning an empty array
- Filtered out nil/blank educational institute values before returning results

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

## Summary by CodeRabbit

* **Bug Fixes**
  * Typeahead now trims input, enforces a 2-character minimum, safely handles special characters, excludes empty/null names, and returns empty results for invalid queries or timeouts.

* **Performance**
  * Search results are cached (short expiry) and result sets are limited to improve response times.

* **Tests**
  * Added tests covering matching, short/blank queries, special characters, and exclusion of nil institute names.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: app/controllers/users/circuitverse_controller.rb [hunk 1]
```diff
@@ -46,6 +45,24 @@ def groups
...
   private
 
+    def fetch_educational_institutes(query)
+      sanitized_query = query.gsub(/[^a-zA-Z0-9\s]/, "")
+      return [] if sanitized_query.blank?
+
+      # Use full-text search with existing GIN-indexed tsvector column
+      # This is much faster than LIKE with leading wildcard
+      Rails.cache.fetch("typeahead_institutes/#{sanitized_query.downcase}", expires_in: 5.minutes) do
+        User.where("searchable @@ plainto_tsquery('english', ?)", sanitized_query)
```
[the remark this answers is not in the document]
Comment (ayuxsh009) [reply app/controllers/users/circuitverse_controller.rb:67]: Fixed — added the 2-char check after sanitization to avoid short queries hitting the DB. Thanks!

**Comment by naman79820** [comment]
closing as not planned
